### PR TITLE
Add fixups from upstream schnorrsig PR

### PR DESCRIPTION
**Pull request by jonasnick**
f make helper functions static
f hash noncedata into nonce in nonce_function_bipschnorr
f expose nonce_function_bipschnorr
f fix undefined behavior when shifting an int 31 places
f add cplusplus ifdef to schnorrsig include file
f hash complete pubkey into batch seed

EDIT:
f chacha20 for bigendians
f add schnorrsig to travis

### CODE REVIEW

**Comment by jonasnick** [comment]
rebased

File: .travis.yml [hunk 1]
```diff
@@ -11,22 +11,22 @@ cache:
...
-    - FIELD=auto  BIGNUM=auto  SCALAR=auto  ENDOMORPHISM=no  STATICPRECOMPUTATION=yes  ASM=no  BUILD=check  EXTRAFLAGS=  HOST=  ECDH=no  RECOVERY=no  EXPERIMENTAL=no  JNI=no GENERATOR=no RANGEPROOF=no WHITELIST=no
+    - FIELD=auto  BIGNUM=auto  SCALAR=auto  ENDOMORPHISM=no  STATICPRECOMPUTATION=yes  ASM=no  BUILD=check  EXTRAFLAGS=  HOST=  ECDH=no  RECOVERY=no  EXPERIMENTAL=no  JNI=no GENERATOR=no RANGEPROOF=no WHITELIST=no SCHNORRSIG=no
     - GUAVA_URL=https://search.maven.org/remotecontent?filepath=com/google/guava/guava/18.0/guava-18.0.jar GUAVA_JAR=src/java/guava/guava-18.0.jar
   matrix:
-    - SCALAR=32bit    FIELD=32bit       EXPERIMENTAL=yes RANGEPROOF=yes WHITELIST=yes GENERATOR=yes
-    - FIELD=64bit     EXPERIMENTAL=yes RANGEPROOF=yes WHITELIST=yes GENERATOR=yes
+    - SCALAR=32bit    FIELD=32bit       EXPERIMENTAL=yes RANGEPROOF=yes WHITELIST=yes GENERATOR=yes SCHNORRSIG=yes
+    - FIELD=64bit     EXPERIMENTAL=yes RANGEPROOF=yes WHITELIST=yes GENERATOR=yes  SCHNORRSIG=yes
     - SCALAR=32bit    RECOVERY=yes
-    - SCALAR=32bit    FIELD=32bit       ECDH=yes  EXPERIMENTAL=yes
```
Comment (real-or-random) [inline .travis.yml]: This change makes Travis fail (and I guess this change is just an accident).

**Comment by jonasnick** [comment]
fixed

**Comment by real-or-random** [comment]
ACK read the code, tested it

can you squash?

**Comment by real-or-random** [comment]
and feel free to add https://github.com/bitcoin-core/secp256k1/pull/558/commits/1f5ca3945e2d7a87db81f751a23e4092418e3ebf

**Comment by jonasnick** [comment]
Cherry picked commit and squashed.

**Pull request merged**